### DISCUSSION TITLE
Message bubbling

### DISCUSSION THREAD

**Discussion by zmerp** [hecrj/iced#1105] [Q&A] [answered]
I'm making an app where UI controls are created and placed programmatically from a schema, in a tree fashion arbitrarily deep. I structured the UI into pseudo-widgets, structs where each one has a view() and update() method. The UI is connected to a server which is in charge of processing messages, which can change the state of the UI. For example, when the user clicks on a switch, a message is bubbled up and sent to the server; the server responds with its new internal state that refreshes all the UI.

The problem I have is that handling simple updates that require fast visual feedback and no interaction with the server is cumbersome. Each child widget needs to bubble up its message and then it has to be re-routed to the child to reach its update method(), and to do so I need to build a trace of the route. What I would like is custom widget support where each widget can have its own state and can manage its own messages without bubbling up and down the messages. Is there any way of achieving this?

**Comment by hecrj** [comment, marked as answer]
> I structured the UI into pseudo-widgets, structs where each one has a view() and update() method.

Why? The `Message`, `update`, `view` triplet introduces the most complexity, you should never use it as a default to design your UI. Instead, you should start with simple view helpers and flat state first!

Overall, you should hand-craft the simplest API for the specific use cases of each part of your app. Avoid applying the same design pattern everywhere! If all you have is a hammer, everything looks like a nail.

**Reply by zmerp** [reply]
The flat state is probably the best approach since each control is independent from its parent and their children. I can put the state of each control in a vector and reference it using indices. Thank you 🙂

**Reply by zmerp** [reply]
Unfortunately this turned out to be a lot more convoluted because of indices hell and even impossible because of lifetimes. I cannot put every control in a flat vector, because I cannot mutably borrow it multiple times. I need to do so because I need to borrow (button) states of different controls at different tree levels in a single `Element<'a, _, _>`. It seems the solution was really to let each control manage its children. Each child control makes sure to mutably borrow its whole subtree at once. But this also means I have to make each control bubble up and down the children messages.

**Reply by zmerp** [reply]
Another solution is to make sure all the controls are constructed in a single flat for-loop, and each control in broken into stateful part + container for the children; but this way I cannot let each control conditionally render its children. At this point I would break encapsulation so much it becomes a mess. I have 12 types of controls, of which 7 can contain children.

**Reply by zmerp** [reply]
Just a heads-up. I'm having success on the "pseudo-widgets" route, even though verbose because of the construction of the path to route the messages. To avoid the proliferation of message types and casting, I'm using a single message type for each control, and unused variants match to `unreachable!()`. The root control needs to align each descendant control, and to achieve this I'm making each pseudo-view method return a single `DrawingResult<'a>` with label + control, and so I solve the double mutable borrow.
